Initial support for elliptical curves

melowe opened the pull request:
1. Refactored encryption and config modules to allow support for alternative encryptors to NACL. 
2. Modified test fixture not to assume NACL. 
3. Refactored key pair loading to facilitate injection of encryptor.
4. Moved new keypair password generation out of config module as this is essentially handling cli command/actions into the config cli module. 

Cli changes
`tessera -updatepassword `requires `--encryptor.type=[NACL|EC]` option
`tessera -keygen` requires `--encryptor.type=[NACL|EC] `option or `{encryptor : {type:NACL or EC }` defining in config 
normal usage default to NACL for backward compatibility.

melowe committed:
Generate key pairs using encryptor rather than fixed values to 
enable tests for different encryptor implementations.

melowe committed:
Reapplied changes made in initial feature/encryption_alternative_curves
 branch.

melowe committed:
ignore push test as fixtures are tricky to generate and this is covered 
in our send scenerios. Add todo comments to failure scenerios documentating 
 the danger of false positives.

melowe committed:
Add initial AEC encryptor module, pending config integration.

melowe committed:
Ensure that JACL encryptor is returned despite AEC being available in 
classloader. Temporary measure until config changes for alternative 
encryptors is added. 

Conflicts:
	encryption/encryption-api/src/main/java/com/quorum/tessera/encryption/EncryptorFactory.java

melowe committed:
Move payload build to be inner class to control and tighten up creation 
logic. Mock payloads in test in code module reducing cross impacts 
of any changes to encoding logic. 

Conflicts:
	enclave/enclave-api/src/main/java/com/quorum/tessera/enclave/EnclaveImpl.java
	enclave/enclave-api/src/main/java/com/quorum/tessera/enclave/EncodedPayload.java
	enclave/enclave-api/src/main/java/com/quorum/tessera/enclave/EncodedPayloadBuilder.java
	enclave/enclave-api/src/main/java/com/quorum/tessera/enclave/PayloadEncoderImpl.java
	enclave/enclave-api/src/test/java/com/quorum/tessera/enclave/EnclaveTest.java
	enclave/enclave-api/src/test/java/com/quorum/tessera/enclave/PayloadEncoderTest.java
	enclave/enclave-jaxrs/src/test/java/com/quorum/tessera/enclave/rest/Fixtures.java
	tessera-core/src/test/java/com/quorum/tessera/transaction/ResendManagerTest.java
	tessera-sync/src/test/java/com/quorum/tessera/sync/Fixtures.java

melowe committed:
Add support for encryptor configuration.

melowe committed:
Move config properties into root package to stop moxy generating 
unwanted namespaces.

melowe committed:
Change type to enumeration and not string

melowe committed:
Add type property to factories and filter on that. Allow optional 
injection of properties as map for custom configuration when creating 
encryptors. 

Conflicts:
	encryption/encryption-api/src/main/java/com/quorum/tessera/encryption/EncryptorFactory.java
	encryption/encryption-api/src/test/java/com/quorum/tessera/encryption/MockEncryptorFactory.java
	encryption/encryption-jnacl/src/main/java/com/quorum/tessera/nacl/jnacl/JnaclFactory.java
	encryption/encryption-kalium/src/main/java/com/quorum/tessera/nacl/kalium/KaliumFactory.java

melowe committed:
All code changes required to add alternative encryptor implementations, 
specifically elliptical curves. Changed test fixtures across project not to
assume nacl. 

Conflicts:
	config-migration/src/main/java/com/quorum/tessera/config/builder/ConfigBuilder.java
	config/src/main/java/com/quorum/tessera/config/keypairs/InlineKeypair.java
	config/src/main/java/com/quorum/tessera/config/keys/KeyEncryptorFactory.java
	enclave/enclave-api/src/main/java/com/quorum/tessera/enclave/EnclaveFactory.java
	encryption/encryption-api/src/test/java/com/quorum/tessera/encryption/EncryptorFactoryTest.java
	key-generation/src/main/java/com/quorum/tessera/key/generation/DefaultKeyGeneratorFactory.java
	key-generation/src/main/java/com/quorum/tessera/key/generation/FileKeyGenerator.java
	tests/acceptance-test/pom.xml
	tests/acceptance-test/src/test/java/com/quorum/tessera/test/P2pTestSuite.java
	tests/acceptance-test/src/test/java/com/quorum/tessera/test/cli/keygen/FileKeygenSteps.java
	tests/acceptance-test/src/test/java/com/quorum/tessera/test/grpc/GrpcSuite.java
	tests/acceptance-test/src/test/java/com/quorum/tessera/test/rest/RestSuite.java
	tests/acceptance-test/src/test/java/com/quorum/tessera/test/rest/RestSuiteSimple.java
	tests/acceptance-test/src/test/java/com/quorum/tessera/test/rest/SendWithRemoteEnclaveReconnectIT.java
	tests/acceptance-test/src/test/java/com/quorum/tessera/test/rest/StressRestSuite.java
	tests/acceptance-test/src/test/java/com/quorum/tessera/test/ws/RestSuiteP2pWebsocketH2.java
	tests/acceptance-test/src/test/java/config/ConfigBuilder.java
	tests/acceptance-test/src/test/java/suite/ExecutionContext.java
	tests/acceptance-test/src/test/java/suite/ProcessConfig.java
	tests/acceptance-test/src/test/java/suite/ProcessConfiguration.java
	tests/acceptance-test/src/test/java/suite/TestSuite.java

melowe committed:
remove commented code

melowe committed:
Rename aec to ec

melowe committed:
avoid null pointer

melowe committed:
Dont provide default encrptor config intenally. Require that config
has config defined.

melowe committed:
Refactor config processing . 
Move password file generation out of config factory.

melowe committed:
Remove catch block used for debugging

melowe committed:
explicitly define passwords list as transient to prevent it 
being marshalled.

melowe committed:
Don't store passwords to file system prefer to store in password list 
to be consumed by password resolver. Add tests to ensure that implicit
protocol using absence or presence of passwords or passwordFile elements
are still working as intended.

namtruong committed:
Inject the encryptor into FileSystemKeyPair instead of InlineKeyPair
Fix some acceptance tests

melowe committed:
Validate that updatepassword requires that the encrpytor type is defined. 
Keygen requires encrpytor type in config or as arg. 
For normal running default encryptor to NACL

melowe committed:
Remove recursive field check as cant compare different encryptor 
implementations.

melowe committed:
Catch any error and log during enclave creation.

melowe committed:
Dont validate key presents after unmarshal as this is suppressed unless 
a key encryptor is present.

melowe committed:
Add encryptpr config to config object to deal with cases where not 
defined on original config file.

melowe committed:
dont sys out config

melowe committed:
Check config isnt null before setting encryptor.

melowe committed:
Add test for log and throw exception handling.

melowe committed:
Remove constructer that adds default encryptor as defaults are 
provided in too many locations.

melowe committed:
Add hashcode and equals.

melowe committed:
extend Configitem to add recursive equals and hashcode

melowe committed:
Remove whitespace, no fucntional changes,.

melowe committed:
Merge branch 'master' into encryption_alt_curves

melowe committed:
Merge branch 'master' into encryption_alt_curves

melowe committed:
Merge branch 'master' into encryption_alt_curves

namtruong commented on cli/config-cli/src/main/java/com/quorum/tessera/config/cli/DefaultCliAdapter.java:
```diff
@@ -138,24 +148,31 @@ public CliResult execute(String... args) throws Exception {
 
-            //return early so other options don't get processed
-            return null;
+        if (!commandLine.hasOption("configfile") && !commandLine.hasOption("keygen")) {
```
Need to add `&& !commandLine.hasOption("updatepassword")`

melowe committed:
Default encryptor type to NACL. Ensure that one of configfile, uddatepassword
or keygen is defined on Cli.

melowe committed:
Set JSON_REDUCE_ANY_ARRAYS on marshaller to stop map marshal returning 
an array of values. Add tests for marshal and unmarshal. Ensure that
w3c element is also handled in xml adapter as unmarhsal receives a list 
of dom elements rather than jaxb elements.
